feat: //FILES now supports globbing and new --resources option

quintesse opened the pull request:
The new `--resources` (`-r`) option allows the user to set additional
files for the script (`--files` wasn't available anymore).

Fixes #1338

quintesse commented on src/test/java/dev/jbang/cli/TestRun.java:
```diff
@@ -8,17 +8,7 @@
-import static org.hamcrest.Matchers.matchesPattern;
-import static org.hamcrest.Matchers.not;
-import static org.hamcrest.Matchers.startsWith;
+import static org.hamcrest.Matchers.*;
```
Not sure why this happens. Shouldn't the `spotlessApply` expand this automatically?

maxandersen replied:
no, spotless apply doesn't unroll import.

quintesse commented:
So perhaps the only thing I'm not 100% happy with is the fact that `-f` (and `--file`, singular) are already taken so I had to use `--resources`. The `-r` was also already taken, but as a shortcut for `--jfr` which I am assuming isn't much used anyway so a shortcut isn't really necessary.

But of course you might disagree. Which would either mean not giving `--resources` a shortcut, or try to rename the current `--file`/`-f`, or come up with some other name besides `--resources`. I'm open to suggestions :-)

maxandersen commented:
-r for --jfr we can fix/change. 

Where are we using --file? (On mobile atm)

maxandersen commented:
`--file/-f` is for alias/catalogs/config not build/run/export etc.

`--files` would be okey would it not? can you imagine there would be conflicts?

quintesse commented:
--file/-f is is for _all_ commands because the config is always read, regardless of the command, so it's part of the base/root JBang command,

Now could/should that have been called `--config` instead of `--file`? Yes, perhaps, but is that something we can change now? That seems more "dangerous" (ie potentially breaking people's scripts) than taking away `-r` from `--jfr`.

quintesse commented:
Hmm wait, that _is_ actually called `--config`, so what was the conflict I was getting yesterday? ... let me check...

quintesse commented:
Ok, that's weird I just changed it to `--files` and `-f` and things just worked... I'm sure I got conflicts yesterday! Anyway, added a commit to the PR for that. Much nicer this way. Internally the code still calls them "resources", but that's in line with existing code, so that's perfectly okay.

maxandersen commented on src/main/java/dev/jbang/cli/Edit.java:
```diff
@@ -46,6 +46,9 @@ public class Edit extends BaseScriptCommand {
 	@CommandLine.Option(names = { "-s", "--sources" }, description = "Add additional sources.")
 	List<String> sources;
 
+	@CommandLine.Option(names = { "-r", "--resources" }, description = "Add additional resources.")
```
should be --files?

maxandersen commented on src/main/java/dev/jbang/cli/BaseBuildCommand.java:
```diff
@@ -20,6 +20,9 @@ public abstract class BaseBuildCommand extends BaseScriptCommand {
 	@CommandLine.Option(names = { "-s", "--sources" }, description = "Add additional sources.")
 	List<String> sources;
 
+	@CommandLine.Option(names = { "-f", "--files" }, description = "Add additional files.")
```
should have "converter=CommaSeparatedConverter.class"

maxandersen commented:
> Ok, that's weird I just changed it to `--files` and `-f` and things just worked... I'm sure I got conflicts yesterday!

because you didn't change it for Alias.java :)

maxandersen commented:
pushed commit that makes "--files" the only way to set this.

Not 100% convinced that is the right way but at least then "-f" does not conflict on Alias.

maxandersen commented:
BUT now ./bootstrap.sh works!

maxandersen commented:
`--files src/main/resources/jbang` gives NPE:

```
[jbang] [ERROR] public class java.lang.NullPointerException
java.lang.NullPointerException
	at dev.jbang.source.SourceSet.lambda$getStableId$1(SourceSet.java:320)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:313)
	at java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:743)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
	at dev.jbang.util.Util.getStableID(Util.java:977)
	at dev.jbang.source.SourceSet.getStableId(SourceSet.java:322)
	at dev.jbang.source.SourceSet.getJarFile(SourceSet.java:312)
	at dev.jbang.source.builders.BaseBuilder.build(BaseBuilder.java:67)
	at dev.jbang.cli.BaseBuildCommand.buildIfNeeded(BaseBuildCommand.java:54)
```

quintesse commented:
> --files src/main/resources/jbang gives NPE:

This is now fixed. In fact it not only doesn't give an error but is treated as if you specified `--files src/main/resources/jbang/**`

maxandersen commented:
> > --files src/main/resources/jbang gives NPE:
> 
> 
> 
> This is now fixed. In fact it not only doesn't give an error but is treated as if you specified `--files src/main/resources/jbang/**`

Great the npe is fixed but is that the right semantics ? The pattern matcher requires normally a /** to trigger deep traverse ?

maxandersen commented:
Note I'm not convinced I'm right here just feel like that is copying too much...

quintesse commented:
Could be, but to me it "feels right". If I'm saying `--sources src/main/java --files src/main/resources` then I probably want to include everything. In fact , in what case would you _not_want that? And then it's easy to just put an `/*` at the end to make it explicit you only want a single directory level.

And I'm not sure if your use of the word "copying" was on purpose, but just to be sure: nothing is being copied here of course, it's just adding paths to the compiler.

maxandersen commented:
Eh? //FILES are copied into the build dir not just added to compiler dirs, right?

quintesse commented:
Ah copied in that way. I thought you meant like sometimes files get copied from their original place to the cache first. But yes, all resource files get copied to the build dir.

maxandersen commented on generated/BuildConfig.java:
```diff
@@ -0,0 +1,9 @@
+package dev.jbang;
```
this shouldn't be committed/added?

quintesse replied:
Oops!

maxandersen reviewed (changes requested):
jbang.jar in root should not be added/committed.

maxandersen commented:
`jbang export portable --files=wonka=notthere/itests/**/*.svg  --force t.java
[jbang] [ERROR] Problem looking for **/*.svg in /Users/max/code/personal/jbangdev/jbang/notthere/itests
[jbang] Run with --verbose for more details
```

is not informative.

it is failing because notthere folder does not exist. Should be stated without having to use --verbose.

quintesse commented:
> is not informative.
> 
> it is failing because notthere folder does not exist. Should be stated without having to use --verbose.
> ```

I'm not saying this couldn't be improved, but no changes were made to this error message, it was already like this before (added by you, no less 😉 ).

quintesse commented:
Ok, added the exception's string tot the message, hopefully that makes it a bit clearer.
